notebooks/11-Advanced-Pandas.ipynb

```diff
@@ -0,0 +1,2864 @@
+    "    \n",
+    "    return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "583c43c1-698a-4566-9516-056e7f1a8602",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "Something to note here is how much extra work we're doing to apply a function in parallel:\n",
+    "\n",
+    "* Split the data frame into N copies\n",
+    "* Fork 4 new Python processes\n",
+    "* Run the function N times\n",
+    "* Concatenate the results back into one data frame\n",
+    "* Clean up"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "033d7990-8f94-4461-b7e7-6be899bb4f3e",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "source": [
+    "In other words: you should only do this if the code you're parallelizing is sufficiently time-intensive to make it worthwhile. Otherwise you could easily end up taking more time than you would if you'd remained single-threaded."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c5027eb8-bc68-4b4f-bb27-3b0f3f1a1cbc",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "**Note:** In general, if you have a substantial amount of data and wish to leverage multiple cores or CPUs, you might be better served investigating a library that is built from the ground up to allow for distributed processing. Dask and PySpark are two possible alternatives in this space."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "932e9eec-e941-455b-9824-42d1827e55bd",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "## Joining Data Frames"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6c7e7b8e-cd33-459e-a9f7-2586f1de56b1",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "source": [
+    "It's quite common to have mulitple data frames that need to be joined together. Consider the example of one \"fact\" table and multiple \"dimension\" tables, all stored separately:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "65fd9fda-a979-432e-9428-97d386e37e21",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "fact = (\n",
+    "    pd.DataFrame({\n",
+    "        'dim_id': [1, 2, 3], \n",
+    "        'val_x': ['x1', 'x2', 'x3']\n",
+    "    })\n",
+    ")\n",
+    "dim = (\n",
+    "    pd.DataFrame({\n",
+    "        'dim_id': [1, 2, 4], \n",
+    "        'val_y': ['y1', 'y2', 'y4']\n",
+    "    })\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6500446b-2240-4d51-b875-604b7a5021d7",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "source": [
+    "We have several different options:"
+   ]
+  },
```

Comment:
I feel like this fragment is a little unnecessary, at least hanging at the bottom of this slide.